src/edu/washington/escience/myria/operator/agg/Aggregate.java

```diff
@@ -103,14 +110,29 @@ protected TupleBatch fetchNextReady() throws DbException {
    * @return A batch's worth of result tuples from this aggregate.
    * @throws DbException if there is an error.
    */
-  private void generateResult() throws DbException {
+  protected void generateResult() throws DbException {
+    Schema inputSchema = getChild().getSchema();
     for (TupleBatch tb : groupStates.getData().getAll()) {
       List<Column<?>> columns = new ArrayList<Column<?>>();
       columns.addAll(tb.getDataColumns().subList(0, gfields.length));
-      for (Aggregator agg : emitAggs) {
-        columns.addAll(agg.emitOutput(tb));
+      int stateOffset = gfields.length;
+      int emitOffset = 0;
+      for (AggregatorFactory factory : factories) {
+        int stateSize = factory.generateStateSchema(inputSchema).numColumns();
+        int emitSize = factory.generateSchema(inputSchema).numColumns();
+        TupleBatch state =
+            tb.selectColumns(MyriaArrayUtils.range(stateOffset, stateOffset + stateSize));
+        for (GenericEvaluator eval : emitEvals.subList(emitOffset, emitOffset + emitSize)) {
+          columns.addAll(eval.evaluateColumn(state).getResultColumns());
```

Comment:
Yep I also have realized it. I don't know if we can assume that `Aggregate` would only have singe-valued emit expressions? For now it seems to be true, even when Python UDFs are involved. Will check with @parmitam .